tracpro/contacts/fields.py

```diff
@@ -41,14 +44,75 @@ def render(self, name, value, attrs=None):
         return mark_safe(''.join(output))
 
 
+def validate_twitter_handle(handle):
+    handle_regex = r'^[a-zA-Z0-9_]{1,15}$'
+    if not re.match(handle_regex, handle):
+        raise ValidationError(
+            _("That is not a valid format Twitter handle.  A valid handle has only "
+              "letters A-Z, numbers 0-9, and underscores (_), and has up to 15 "
+              "characters."))
+
+
+def validate_numeric_twitter_id(numeric_id):
+    if not numeric_id.isdigit():  # True only if all chars are digits
+        raise ValidationError(
+            _("That is not a valid format Twitter ID.  A valid numeric ID has only "
+              "numbers 0-9."))
+
+
+def validate_twitter_id(twitter_id):
+    if '#' in twitter_id:
+        numeric_id, handle = twitter_id.split('#', 1)
+        validate_twitter_handle(handle)
+        validate_numeric_twitter_id(numeric_id)
+    else:
+        raise ValidationError(
+            _("That is not a valid format Twitter ID.  The correct format is: "
+              "<numeric id>#<Twitter handle>"))
+
+
+def validate_phone_used(number):
+    from .models import Contact
+    contacts = Contact.objects.all()
```

Comment:
To see if a number is in use, it's probably more efficient to just query for contacts with that phone number, e.g. ``Contact.objects.filter(urn="tel:%s" % number)``.

However, we don't really know here whether the number we're validating belongs to a contact that already exists. If so, it's okay for the number to be in use, because it's in use by the contact we're validating it for.  Typically we'd handle this by excluding the pk of the existing contact from the query, so we'll have to pass that along from the ``clean`` method if the record has a pk.  For that matter, the ``clean`` method is on the field class, and I don't know offhand whether it even has access to the record it belongs to.  You'd hope so, of course.

Actually there's another problem - we should be ignoring any punctuation in the number, since the same number could be written different ways. I'm not really sure how to make that efficient, except maybe to store another field with the canonical form of the urn, with punctuation stripped or something, and use that to check for dupes. This gets more and more messy, unfortunately.

Which leads me to wonder if it's really worth trying to avoid duplicates here, if it's going to be this much work. Do we know if RapidPro allows contacts with the same phone number?

Reply:
+1 on filtering and excluding(when Updating).  I left this in a state that 'worked' but then realized that this might be a little more involved than I have here, I will discuss below.*

Yes, I think I have a solution for the telephone number having  

* Rapidpro doesn't allow duplicate phone numbers, but this is actually for the whole org, not just the contacts that TracPro has synced based on Panels and Cohorts.  When you try to sync the new contact (that you were able to save on TracPro given that the number was unique for all TracPro contacts), the sync still fails in the background with an error message not visible to a user (error: tracpro/contacts/utils --> line 187).  I'm deciding what to do here.